File: java/src/main/java/problems101_200/BestTimeToBuyAndSellStock2.java
package problems101_200;

enum BestTimeToBuyAndSellStock2 {
    DEFAULT {
        @Override
        int maxProfit(int[] prices) {
            int profit = 0;
            boolean hold = false;
            for (int i = 0; i < prices.length; i++) {
                if (i != prices.length - 1 && prices[i + 1] > prices[i]) {
                    if (!hold) {
                        hold = true;
                        profit -= prices[i];
                    }
                    continue;
                }
                if (!hold) {
                    continue;
                }
                hold = false;
                profit += prices[i];
            }
            return profit;
        }
    };

    abstract int maxProfit(int[] prices);
}

File: kotlin/src/main/kotlin/problems101_200/BestTimeToBuyAndSellStock2.kt
package problems101_200

enum class BestTimeToBuyAndSellStock2 {
    DEFAULT {
        override fun maxProfit(prices: IntArray): Int {
            var profit = 0
            var hold = false
            for (i in prices.indices) {
                if (i != prices.size - 1 && prices[i + 1] > prices[i]) {
                    if (!hold) {
                        hold = true
                        profit -= prices[i]
                    }
                    continue
                }
                if (!hold) {
                    continue
                }
                hold = false
                profit += prices[i]
            }
            return profit
        }
    },
    ;

    abstract fun maxProfit(prices: IntArray): Int
}
